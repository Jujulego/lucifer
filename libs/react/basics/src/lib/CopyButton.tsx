import React, { FC, MouseEvent, useEffect, useState } from 'react';
import copy from 'copy-to-clipboard';

import {
  IconButton, Tooltip,
  IconButtonProps
} from '@material-ui/core';
import FileCopyIcon from '@material-ui/icons/FileCopy';

// Types
export interface CopyButtonProps extends IconButtonProps {
  /** Content to be copied */
  text: string;

  /**
   * Mime type of content
   * @default text/plain
   */
  format?: string;

  /**
   * Tooltip content
   * @default Copié !
   */
  tooltip?: string;

  /**
   * Tooltip duration (in milliseconds)
   * @default 1500
   */
  tooltipTimeout?: number;

  /** Copied callback, called after a successful copy */
  onCopied?: () => void;
}

// Component
const CopyButton: FC<CopyButtonProps> = (props) => {
  // Props
  const {
    text, format = "text/plain",
    tooltip = "Copié !",
    tooltipTimeout = 1500,
    onCopied, onClick,
    ...btn
  } = props;

  // State
  const [showTooltip, setShowTooltip] = useState(false);

  // Handler
  const handleCopy = (event: MouseEvent<HTMLButtonElement>) => {
    if (onClick) onClick(event);

    // Copy
    const success = copy(text, { format });

    if (success) {
      setShowTooltip(true);
      if (onCopied) onCopied();
    }
  };

  // Effect
  useEffect(() => {
    if (showTooltip) {
      const id = setTimeout(() => setShowTooltip(false), tooltipTimeout);
      return () => clearTimeout(id);
    }
  }, [showTooltip, tooltipTimeout]);

  // Render
  return (
    <Tooltip
      title={tooltip} open={showTooltip}
      onOpen={() => setShowTooltip(true)}
      onClose={() => setShowTooltip(false)}
      disableFocusListener disableHoverListener disableTouchListener
    >
      <IconButton {...btn} onClick={handleCopy}>
        <FileCopyIcon fontSize="small" />
      </IconButton>
    </Tooltip>
  );
};

export default CopyButton;
