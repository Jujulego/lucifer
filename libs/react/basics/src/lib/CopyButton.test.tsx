import React from 'react';
import copy from 'copy-to-clipboard';
import { act, screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CopyButton from '../lib/CopyButton';

// Mocks
jest.mock('copy-to-clipboard');

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  (copy as jest.Mock).mockReset();
});

// Tests
describe('CopyButton', () => {
  it('should render a button', () => {
    // Render
    render(
      <CopyButton text="test" />
    );

    // Check elements
    expect(screen.queryByRole('button')).toBeInTheDocument();
  });

  it('should copy on click', () => {
    const spy = jest.fn();
    (copy as jest.Mock).mockImplementation(() => true);

    // Render
    render(
      <CopyButton text="test" tooltip="tooltip" onCopied={spy} />
    );

    // Interact
    const btn = screen.getByRole('button');
    userEvent.click(btn);

    // Check callbacks
    expect(spy).toBeCalled();
    expect(copy).toHaveBeenCalledWith('test', { format: 'text/plain' });

    // Check tooltip
    const tooltip = screen.getByRole('tooltip');
    expect(tooltip).toHaveTextContent('tooltip');

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
  });
});
