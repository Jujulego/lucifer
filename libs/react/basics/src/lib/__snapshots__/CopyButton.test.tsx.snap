// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<WithStyles(ForwardRef(Tooltip))
  disableFocusListener={true}
  disableHoverListener={true}
  disableTouchListener={true}
  open={false}
  title="CopiÃ© !"
>
  <WithStyles(ForwardRef(IconButton))
    onClick={[Function]}
  >
    <Memo([object Object])
      fontSize="small"
    />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Tooltip))>
`;
